internal/datastore/common/revisions.go

```diff
@@ -0,0 +1,89 @@
+package common
+
+import (
+	"context"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/shopspring/decimal"
+
+	"github.com/authzed/spicedb/internal/datastore"
+)
+
+type RemoteNowFunction func(context.Context) (datastore.Revision, error)
+
+type RemoteClockRevisions struct {
+	QuantizationNanos      int64
+	GCWindowNanos          int64
+	FollowerReadDelayNanos int64
+	MaxRevisionStaleness   time.Duration
+	NowFunc                RemoteNowFunction
+
+	lastQuantizedRevision decimal.Decimal
+	revisionValidThrough  time.Time
+}
+
+func (rcr *RemoteClockRevisions) OptimizedRevision(ctx context.Context) (datastore.Revision, error) {
+	// ctx, span := tracer.Start(ctx, "OptimizedRevision")
+	// defer span.End()
+
+	localNow := time.Now()
+	if localNow.Before(rcr.revisionValidThrough) {
+		log.Ctx(ctx).Debug().Time("now", localNow).Time("valid", rcr.revisionValidThrough).Msg("returning cached revision")
+		return rcr.lastQuantizedRevision, nil
+	}
+
+	log.Ctx(ctx).Debug().Time("now", localNow).Time("valid", rcr.revisionValidThrough).Msg("computing new revision")
+
+	nowHLC, err := rcr.NowFunc(ctx)
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+
+	// Round the revision down to the nearest quantization
+	// Apply a delay to enable follower reads: https://www.cockroachlabs.com/docs/stable/follower-reads.html
```

Comment:
Does this comment still apply in the common code?

Reply:
The delay is still there and still configurable, even if it's only useful for CRDB. Maybe we'll find other uses for it in the future?

Reply:
We might want to add text saying the CRDB is an example